pkg/kubectl/cmd/exec.go

```diff
@@ -174,16 +174,28 @@ func (p *ExecOptions) Run() error {
 		return err
 	}
 
-	if pod.Status.Phase != api.PodRunning {
-		return fmt.Errorf("pod %s is not running and cannot execute commands; current phase is %s", p.PodName, pod.Status.Phase)
-	}
-
 	containerName := p.ContainerName
 	if len(containerName) == 0 {
 		glog.V(4).Infof("defaulting container name to %s", pod.Spec.Containers[0].Name)
 		containerName = pod.Spec.Containers[0].Name
 	}
 
+	isInitContainer := false
+	for _, initContainer := range pod.Spec.InitContainers {
+		if initContainer.Name == containerName {
+			isInitContainer = true
+			break
+		}
+	}
+
+	if !isInitContainer && pod.Status.Phase != api.PodRunning {
```

Comment:
In the original implementation, we will check the pod status running before we do real exec request, guess we need to remove this status check if we need to support init containers, since pod status is Pending when init containers have not run to completion, also that's why I check the pod status here, exec against init container when the pod is running will fail, seems kubelet will handle this and we need to just get this error in kubectl side.